Bluetooth: Host: cleanup more with bt_disable to allow re-enabling

sean-madigan opened the pull request:
-Reset advertising flags
-Reset conn state
-Reset callbacks
-Reset disconnect handles
-init bt_worq
-Add bt_disable shell command

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

hermabe commented on subsys/bluetooth/host/scan.c:
```diff
@@ -1335,6 +1335,15 @@ void bt_le_scan_cb_unregister(struct bt_le_scan_cb *cb)
+{
+	struct bt_le_scan_cb *cb;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&scan_cbs, cb, node) {
+		sys_slist_find_and_remove(&scan_cbs, &cb->node);
+	}
```
Consider using something like this to avoid iterating over the list unnecessarily:
```suggestion
	while (sys_slist_get(&scan_cbs)) {}
```

hermabe replied:
Forget the thing about extra work. Both versions will remove the first element in a loop.

However, the [docs](https://docs.zephyrproject.org/latest/kernel/data_structures/slist.html#c.SYS_SLIST_FOR_EACH_CONTAINER) for `SYS_SLIST_FOR_EACH_CONTAINER` states:
> Note: the loop is unsafe and thus __cn should not be detached.

sem1-nordic replied:
What does "should not be detached" mean, sorry? So do you reckon `while (sys_slist_get(&scan_cbs)) {}` is the better way to go?

hermabe replied:
I interpret it as "should not be removed from the list". `while (sys_slist_get(&scan_cbs)) {}` is more in line with how we do it for buffers. For the list we could also set the `head` and `tail` pointer to `NULL` to "forget" the list, but then the rest of the list would still exist (the nodes would point to the next nodes still). I am not sure if it matters.

hermabe commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -3480,6 +3485,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(bt_scan_filter_clear_cmds,
 SHELL_STATIC_SUBCMD_SET_CREATE(bt_cmds,
 	SHELL_CMD_ARG(init, NULL, "[no-settings-load], [sync]",
 		      cmd_init, 1, 2),
+	SHELL_CMD_ARG(disable, NULL, HELP_NONE, cmd_disable, 1, 2),
```
There are no optional arguments to the disable command.
```suggestion
	SHELL_CMD_ARG(disable, NULL, HELP_NONE, cmd_disable, 1, 0),
```

hermabe commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2128,6 +2164,11 @@ void bt_conn_cb_register(struct bt_conn_cb *cb)
 
+void bt_conn_cb_unregister(void)
+{
+	memset(&callback_list, 0, sizeof(callback_list));
```
`callback_list` is just a pointer.
```suggestion
	callback_list = NULL;
```

Or maybe go through the list and set the `_next` pointers to `NULL`.

hermabe reviewed (commented):
Nice with some more cleanup! Do we have a way to know if there is anything more that needs to be cleaned up by `bt_disable`? There have been a couple other patches that have added some cleanups that were missing and it would be nice to be sure that we are "complete".

A babblesim test that verifies that enable and disable works would be nice.

We should also consider whether the user-registered `bt_conn_cb` callbacks should not be unregistered. The application is likely to want the same callbacks if/when they re-enable the bluetooth stack.

sem1-nordic commented:
I presume there will be more cleanup required. Currently I have only tested for:
-enable, disable in a loop
-enable, start adv, disable in a loop
-enable, start scan, disable in a loop
-enable, connect, disable in a loop

hermabe commented:
> I presume there will be more cleanup required. Currently I have only tested for:
> - enable, disable in a loop
> - enable, start adv, disable in a loop
> - enable, start scan, disable in a loop
> - enable, connect, disable in a loop

That should cover many use cases. However I think we should test something in GATT/ATT and SMP as well.

alwa-nordic commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1096,6 +1096,13 @@ struct bt_le_ext_adv_info {
+ *
+ * Clear all advertising flags.
+ */
+void bt_adv_reset_adv_flags(void);
```
Don't put internal functions in the public API headers in `/include/zephyr/bluetooth`.

alwa-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -671,6 +671,37 @@ static bool send_buf(struct bt_conn *conn, struct net_buf *buf)
+
+	bt_conn_reset_rx_state(conn);
+
+	k_work_reschedule(&conn->deferred_work, K_NO_WAIT);
```
Does it matter if `k_work_reschedule` returns an error here? If no, can you add a comment in the code explaining that we are ignoring the errors on purpose (and why it's not a problem)? Same questions for the addition of `k_work_cancel_delayable` further down.

jori-nordic replied:
this seems super similar to `conn_cleanup(struct bt_conn *conn)`. I think this fn can just be a call to that + the `set_state` after.

alwa-nordic commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3778,10 +3788,30 @@ int bt_disable(void)
+	bt_pub_key_callback_clear();
+	bt_conn_cleanup_all_conns();
+	disconnected_handles_reset();
+	bt_conn_cb_unregister();
```
Please add a sentence to `bt_conn_cb_register` API doc saying that the callbacks registered trough that function are automatically cleared by `bt_disable` after all connections are destroyed and all relevant callbacks have been invoked.

RFC: I suggest we don't clear the conn_cb listeners as part of `bt_disable`. It feels cleaner to make the listener registration independent. To further make it independent, we can explicitly allow registering before `bt_enable`. We can provide a `bt_conn_cb_unregister(struct bt_conn_cb *)` API.

hermabe replied:
Agree with the RFC.

sem1-nordic replied:
Same with the scan_cbs and pub_key_cb_slist, do you reckon?

jori-nordic replied:
I don't really agree with the RFC if the role of bt_disable is to perform a soft-reset of the whole bt subsystem. 
In that case, some state would linger on (in the form of those callbacks being registered). It also makes the mental model of what's happening clearer and forces the app users to not depend on said state.

I don't think a lot of users are currently calling `bt_disable` (we'd have more bug reports), so the few that do it can do the extra work of re-registering any callbacks they depend on.

alwa-nordic replied:
> Same with the scan_cbs and pub_key_cb_slist, do you reckon?

`scan_cbs` is the same, yes.

`pub_key_cb_slist` actually contains callbacks as opposed to event listeners. The callback entries are waiting for some operation to complete, and must all be invoked/completed exactly once. You can see in `bt_hci_evt_le_pkey_complete` that the operation can fail and the callback will then invoked with NULL. Instead of clearing this list manually, the operations that the callbacks are waiting on should be aborted and failed.

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -991,6 +991,18 @@ struct bt_conn_cb {
+ *
+ *  Unregister callbacks to cleanup connections.
+ */
+void bt_conn_cb_unregister(void);
```
I think it should be <code>bt_conn_cb_unregister<b>_all</b></code>. The plain `bt_conn_cb_unregister` must be reserved for a future `bt_conn_cb_unregister(struct bt_conn_cb *cb)`, that is the symmetrical API to `bt_conn_cb_register`.

jori-nordic commented:
> I presume there will be more cleanup required. Currently I have only tested for: -enable, disable in a loop -enable, start adv, disable in a loop -enable, start scan, disable in a loop -enable, connect, disable in a loop

Could you make a bsim test for `bt_disable`? And add these cases:
- the ones you mention having tested with (in the shell I suppose)
- a case for reproducing https://github.com/zephyrproject-rtos/zephyr/issues/48304 . make sure to check if it was indeed broken before your changes
- the case @rugeGerritsen suggests [here](https://github.com/zephyrproject-rtos/zephyr/issues/48304#issuecomment-1202059928)

jori-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -991,9 +1022,14 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 			break;
+		case BT_CONN_DISCONNECTED:
+			/* Can only happen if bt_disable is called whilst
+			 * in a connections.
```
connection

jori-nordic commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -103,6 +103,21 @@ int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
 	return 0;
 }
 
+void bt_pub_key_callback_clear(void)
```
either callback**s** or `bt_pub_key_cb_clear_all(void)`

jori-nordic commented on subsys/bluetooth/host/adv.c:
```diff
@@ -269,6 +269,17 @@ void bt_le_ext_adv_foreach(void (*func)(struct bt_le_ext_adv *adv, void *data),
+{
+#if defined(CONFIG_BT_EXT_ADV)
+	for (size_t i = 0; i < ARRAY_SIZE(adv_pool); i++) {
+		memset(adv_pool[i].flags, 0, BT_ADV_NUM_FLAGS);
```
I'm wondering if we should just memset the whole `struct bt_le_ext_adv` element instead of just the `flags` member.
I'd prefer if we did this, if it doesn't break anything.
@alwa-nordic thoughts?

alwa-nordic replied:
I agree. The less the code has to "know", the better. Memseting the whole `adv_pool` variable is simplest. I have not verified that it's safe to do so, though.

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -1335,6 +1335,15 @@ void bt_le_scan_cb_unregister(struct bt_le_scan_cb *cb)
 	sys_slist_find_and_remove(&scan_cbs, &cb->node);
 }
 
+void bt_le_scan_uninit(void)
```
considering we're just removing callbacks here, could you update the name to reflect that and be consistent with the other fns you added to remove callbacks?

jori-nordic reviewed (changes requested):
Thanks for the PR! Some changes highlighted.
Please consider making a bsim test (see [my previous comment](https://github.com/zephyrproject-rtos/zephyr/pull/50046#issuecomment-1246407878)); I can assist you with this.

jori-nordic commented:
> > I presume there will be more cleanup required. Currently I have only tested for: -enable, disable in a loop -enable, start adv, disable in a loop -enable, start scan, disable in a loop -enable, connect, disable in a loop
> 
> Could you make a bsim test for `bt_disable`? And add these cases:
> 
> * the ones you mention having tested with (in the shell I suppose)
> * a case for reproducing [bt_disable() does not work properly on nRF52 #48304](https://github.com/zephyrproject-rtos/zephyr/issues/48304) . make sure to check if it was indeed broken before your changes
> * the case @rugeGerritsen suggests [here](https://github.com/zephyrproject-rtos/zephyr/issues/48304#issuecomment-1202059928)

Discussed OOB, bsim test will come in a separate PR.

adrdub commented:
Thanks for this PR, some changes, especially in hci_core.c (in functions bt_enable() and bt_disable()) are crucial for us and would unblock our development with BLE (on stm32wb55) so looking forward to see it merged as soon as possible.

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -269,6 +269,17 @@ void bt_le_ext_adv_foreach(void (*func)(struct bt_le_ext_adv *adv, void *data),
+		memset(&adv_pool[i], 0, sizeof(adv_pool[i]));
+	}
+#else
+	memset(&bt_dev.adv, 0, sizeof(bt_dev.adv));
```
```suggestion
	(void)memset(&bt_dev.adv, 0, sizeof(bt_dev.adv));
```

For the sake of convention

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -698,6 +698,21 @@ static void conn_cleanup(struct bt_conn *conn)
 
+static void conn_destroy(struct bt_conn *conn, void *data)
+{
+	while(sys_slist_get(&conn->tx_pending)) {}
```
```suggestion
	while (sys_slist_get(&conn->tx_pending)) {}
```

Thalley replied:
Also I really dislike that a `pop` function is named `get` :(

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -991,9 +1006,14 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 			 */
 			bt_conn_unref(conn);
 			break;
+		case BT_CONN_DISCONNECTED:
+			/* Can only happen if bt_disable is called whilst
+			 * in a connection.
+			 */
+			bt_conn_unref(conn);
+			break;
```
In line 882-884 we have
```c
if (conn->state == state) {
	BT_WARN("no transition %s", state2str(state));
	return;
}
```

So the previous is correct: We cannot end up in this case

sem1-nordic replied:
Sorry, believe this was meant to go on the line below for `case BT_CONN_CONNECTED`.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -991,7 +991,19 @@ struct bt_conn_cb {
 void bt_conn_cb_register(struct bt_conn_cb *cb);
 
-/**
+/** @brief Unregister all connection callbacks.
+ *
+ *  Unregister all callbacks to cleanup connections.
+ */
+void bt_conn_cb_unregister_all(void);
```
I have a really tough time with this addition. 

We may register the `bt_conn` callbacks from multiple places in the stack, and this allows an application to deregister them all. Not only does this break basically the entire stack, but since there's no callback for the callback being unregistered, there is no way to handle calls to this function. 

Furthermore, this does not handle callbacks registered with `BT_CONN_CB_DEFINE`.

hermabe replied:
Some previous discussion above: https://github.com/zephyrproject-rtos/zephyr/pull/50046#discussion_r969568993

I agree that the callbacks do not need to be unregistered. We should rather supply functions to unregister callbacks and then the application can choose to unregister their own callbacks if they want to.

Thalley replied:
Why do we need to unregister the callbacks at all? 

`bt_conn_cb_register` can called regardless of the BT device state, and I don't see the state of the callbacks as being related to the `bt_disable` or re-enable at all.

hermabe replied:
It is not needed, and since the stack relies on the callbacks they should not be removed.

If however, the application wants to remove them for whatever reason they should do it themselves.

sem1-nordic replied:
Seems to be reasonable to not unregister these callbacks, will update.

Thalley replied:
> It is not needed, and since the stack relies on the callbacks they should not be removed.
> 
> If however, the application wants to remove them for whatever reason they should do it themselves.

Indeed. We haven't had any requests for that functionality AFAIK though. A workaround if you want to unregister is simply to set the callback, in the callback struct, to `NULL`.

sem1-nordic replied:
Would you say the same for `bt_le_scan_unregister_all()`?

Thalley replied:
Yes. Generally I think it's a bad idea to implicit deregister something that an application, or the stack, has registered. None of them would know when or why these are unregistered, and that will generally just break everything (unless we add callbacks for that, but then we are basically back to just leaving the callbacks as is). 

Callbacks are generally not related to the BT stack state.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -3017,6 +3052,7 @@ int bt_conn_init(void)
 	/* Initialize background scan */
 	if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
 		for (i = 0; i < ARRAY_SIZE(acl_conns); i++) {
+			atomic_set(&acl_conns[i].ref, 0);
```
Why is this added? It should not be needed, as the reference count should always be 0 here in the first place.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -719,6 +719,11 @@ int bt_hci_disconnect(uint16_t handle, uint8_t reason)
 static uint16_t disconnected_handles[CONFIG_BT_MAX_CONN];
+static void disconnected_handles_reset(void)
+{
+	memset(disconnected_handles, 0, sizeof(disconnected_handles));
```
```suggestion
	(void)memset(disconnected_handles, 0, sizeof(disconnected_handles));
```

For convention

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1473,6 +1473,11 @@ static int gatt_unregister(struct bt_gatt_service *svc)
 	return 0;
 }
 
+void bt_gatt_unregister_all(void)
+{
+	while (sys_slist_get(&db)) {}
+}
```
I guess it makes sense, but is that necessary? 

If we deregister all the dynamically registered GATT services, then we need to implement a way for the services that are registered to know that they have become unregistered. Many of them have persisting data that needs to be cleared, e.g. `bt_vcs_register` has a `static bool registered` that would prevent the service from being registered again.

sem1-nordic replied:
I was under the idea that we should basically clean everything up from bt_disable, but would be open to the idea of leaving these services registered if that makes more sense

Thalley replied:
This is similar to the callbacks:
It does not really serve a purpose to unregister all dynamically registered GATT service. First of, the statically registered are still there, so this only partially works. Secondly, the services, like the callbacks, can be registered regardless of the BT device state. Lastly, if we do want to unregister them (without any reason that I can think of), then we need to somehow notify whomever registered them in the first place, that they are now unregistered. 

I'm not 100% against this, as semantically it makes sense, but it serves no functional purpose besides just adding additional work to register and unregister, as well as additional work for the developer defining and registering the service, as they also would need to deal with the unregister.

Thalley reviewed (changes requested):
The changes in this PR seemingly makes it impossible to do

```c
bt_enable();
bt_disable();
bt_enable();
```
where the state after `bt_enable()` will be the same. This is due to the PR adding several "unregister" functions for things that are not expected to be unregistered ever.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -991,6 +991,12 @@ struct bt_conn_cb {
  */
 void bt_conn_cb_register(struct bt_conn_cb *cb);
 
+/** @brief Cleanup connections.
+ *
+ *  Reset states of connections and set state to BT_CONN_DISCONNECTED.
+ */
+void bt_conn_cleanup_all_conns(void);
```
What is the reasoning for adding this in the public API? 

It will locally consider all connections as disconnected, but we never send a disconnect to the remote devices, so they will still consider us connected. This seems like a breaking change as well, and I do not see any reason to have this in a public API.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -992,8 +1009,12 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
+			 */
+			break;
 		case BT_CONN_DISCONNECTED:
+		case BT_CONN_DISCONNECTING:
```
Please revert the order of `BT_CONN_DISCONNECTING and `BT_CONN_DISCONNECTED` so that it does not appear as a change

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -992,8 +1009,12 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 			break;
 		case BT_CONN_CONNECTED:
-		case BT_CONN_DISCONNECTING:
+			/* Can only happen if bt_disable is called whilst
+			 * in a connection.
+			 */
```
Or if `bt_conn_cleanup_all_conns` is called by the application at any point :s

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3779,10 +3790,30 @@ int bt_disable(void)
+#endif /* CONFIG_BT_L2CAP_DYNAMIC_CHANNEL */
+#endif /* CONFIG_BT_CONN */
+
+	err = bt_id_delete_all();
+	if (err) {
+		return err;
+	}
```
I understand why it's done like this, but basically we here return an err on `bt_disable`, indicating that it failed, but we have actually disable almost everything at this point. I wonder if this should be replaced with an assert instead? It can't be recovered in any case.

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -90,6 +90,13 @@ static void free_tx_meta_data(struct l2cap_tx_meta_data *data)
 
+void bt_l2cap_free_tx_context(void)
+{
+	while (k_fifo_get(&free_l2cap_tx_meta_data, K_NO_WAIT)) {
+
+	}
```
Do we not get the same result from simply calling `k_fifo_init(&free_l2cap_tx_meta_data);`?

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -785,6 +792,13 @@ static struct bt_l2cap_server *l2cap_server_lookup_psm(uint16_t psm)
 
+void bt_l2cap_server_unregister_all(void)
+{
+	while (sys_slist_get(&servers)) {
+
+	}
```
Do we not get the same result from simply calling `sys_slist_init(&servers);`?

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2714,6 +2728,7 @@ void bt_l2cap_init(void)
 
 #if defined(CONFIG_BT_L2CAP_DYNAMIC_CHANNEL)
 	for (size_t i = 0; i < ARRAY_SIZE(l2cap_tx_meta_data_storage); i++) {
+		memset(&l2cap_tx_meta_data_storage[i], 0, sizeof(l2cap_tx_meta_data_storage[i]));
```
```suggestion
		(void)memset(&l2cap_tx_meta_data_storage[i], 0, 
			     sizeof(l2cap_tx_meta_data_storage[i]));
```

Thalley commented on subsys/bluetooth/host/id.c:
```diff
@@ -1266,6 +1262,32 @@ int bt_id_delete(uint8_t id)
+	return id_delete(id);
+}
+
+int bt_id_delete_all(void)
+{
+	for (uint8_t id = 0U; id < bt_dev.id_count; id++) {
+		int err;
+
+		err = id_delete(id);
+		if (err) {
+			return err;
+		}
+	}
+
+	bt_dev.id_count = 0;
+
+	return 0;
+}
```
This will also delete the `BT_ID_DEFAULT`, which isn't possible via the `bt_id_delete` API. Is this intended? Is the `BT_ID_DEFAULT` added again on `bt_enable`?

sem1-nordic replied:
Yes, `bt_id_init` sets up `BT_ID_DEFAULT` which is called in `bt_enable`

Thalley reviewed (changes requested):
All my major grievances for this PR has been resolved. 

Before accepting it, I want to request a test (e.g. in babblesim) to verify that it is indeed possible to do 
```c
bt_enable();
bt_disable();
bt_enable();
```

jhedberg reviewed (changes requested):
Couple of minor comments. Additionally, please split these changes up into separate commits (with appropriate commit messages), so that each logical change is in its own commit. Thanks.

sean-madigan committed:
Bluetooth: Shell: Add bt_disable to bluetooth shell app

bt_disable can now be used with the bluetooth shell.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

sean-madigan committed:
Bluetooth: Host: init bt_workq and cancel rpa_update work

Init bt_workq in bt_enable so it can be safely disabled.

Cancel rpa_update work upon disable as it is no longer required.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

sean-madigan committed:
Tests: Bluetooth: Add bt_disable babblesim test with gatt

Add babblesim test for bt_disable which enables bt, starts a connection,
GATT discovers, reads, writes then disables BT in a loop.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

r2r0 commented on subsys/bluetooth/host/conn.c:
```diff
@@ -698,6 +698,28 @@ static void conn_cleanup(struct bt_conn *conn)
+
+	bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+
+	k_work_cancel_delayable(&conn->deferred_work);
```
Shouldn't this work be cancelled before other actions performed on `bt_conn`?

sem1-nordic replied:
That makes sense, yes

Thalley reviewed (dismissed):
Thanks for adding the tests :) LGTM

Thalley commented:
@sem1-nordic Will wait with a re-review once the tests are passing :)

sean-madigan committed:
Bluetooth: Host: init fifos and slists before use

In bt_att_init, bt_conn_init and bt_l2cap_init, init the fifos and
slists before use. This means that they can be used after disable
without memory leakage.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

sean-madigan committed:
Bluetooth: Host: Check if EATT l2cap server is registered

bt_eatt_init currently registers a l2cap server, however should not
register a new one if it has already been registered - for example in
bt_enable previously.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

alwa-nordic commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -731,6 +731,11 @@ static int cmd_init(const struct shell *sh, size_t argc, char *argv[])
 
+static int cmd_disable(const struct shell *sh, size_t argc, char *argv[])
+{
+	return bt_disable();
```
It's not correct to pass on the return value. Shell handler functions pick return values from the following table:

https://github.com/zephyrproject-rtos/zephyr/blob/9331591ab1f6a59513ded32d877f62dbf4526b9e/include/zephyr/shell/shell.h#L134-L137

alwa-nordic replied:
I just realized it's not clear what to return when the command is understood, but not successful. @nordic-krch Do you have a suggestion?

sem1-nordic commented:
@cvinayak I am getting an zll assert on the newly added test. It seems this assert was added recently so I didn't see it when writing the test. Would you be able to advise?

> d_00: @00:06:24.760616  ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:476
d_00: @00:06:24.760616  	lll_preempt_calc: Actual EVENT_OVERHEAD_START_US = 127324829
d_00: @00:06:24.760616  @ WEST_TOPDIR/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:476

alwa-nordic commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -731,6 +731,11 @@ static int cmd_init(const struct shell *sh, size_t argc, char *argv[])
 
+static int cmd_disable(const struct shell *sh, size_t argc, char *argv[])
+{
+	return bt_disable();
```
Context: I had deleted my comment here, but it seems @Thalley managed to see it before it was deleted.

I said that shell command handlers should not return errno values, but rather pick from the following list.

https://github.com/zephyrproject-rtos/zephyr/blob/30e4a4b71e3ecf54edef57f8c828a8f487068c1e/include/zephyr/shell/shell.h#L134-L137.

But then I was not so sure since two of choices are negated errno-values.

sean-madigan committed:
Tests: Bluetooth: Add bt_disable babblesim test

Add babblesim test for bt_disable which just enables and disables BT in
a loop.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

sem1-nordic commented:
I have removed the bt_disable with gatt test currently as was hitting zephyr link layer assert. Will be adding more bt_disable babblesim tests in a future task, so will add the bt_disable with gatt test into that task.

Thalley commented:
> I have removed the bt_disable with gatt test currently as was hitting zephyr link layer assert. Will be adding more bt_disable babblesim tests in a future task, so will add the bt_disable with gatt test into that task.

@cvinayak is this fixed by https://github.com/zephyrproject-rtos/zephyr/pull/51709 ?

alwa-nordic commented on subsys/bluetooth/host/id.c:
```diff
@@ -1284,7 +1280,7 @@ int bt_id_delete(uint8_t id)
 		int err;
 
-		err = bt_unpair(id, NULL);
+		err = bt_unpair(id, BT_ADDR_LE_ANY);
```
Why is this changed?

sem1-nordic replied:
Using NULL caused memory issues in some circumstances

sem1-nordic replied:
Fixed the issue now

alwa-nordic commented on subsys/bluetooth/host/id.c:
```diff
@@ -1307,6 +1303,32 @@ int bt_id_delete(uint8_t id)
+	for (uint8_t id = 0U; id < bt_dev.id_count; id++) {
+		int err;
+
+		err = id_delete(id);
+		if (err) {
+			return err;
```
You must ignore `err == -EALREADY`. There may be fragmentation/holes in `bt_dev.id_addr[0:bt_dev.id_count]` if an identity has been deleted before.

I recommend adding a few `id_create` and `id_delete` to the bablesim test. Add two `id_create`, and `id_delete` the first one you created.

sem1-nordic replied:
You make a good point

sem1-nordic replied:
Not sure this works too well in the babblesim test, a unit test would be ideal.

Thalley commented:
> > I have removed the bt_disable with gatt test currently as was hitting zephyr link layer assert. Will be adding more bt_disable babblesim tests in a future task, so will add the bt_disable with gatt test into that task.
> 
> @cvinayak is this fixed by #51709 ?

@sem1-nordic Can you try to re-add the GATT tests? I think the issue you saw might have been fixed.

sem1-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -699,6 +699,29 @@ static void conn_cleanup(struct bt_conn *conn)
+{
+	k_work_cancel_delayable(&conn->deferred_work);
+
+	sys_slist_init(&conn->tx_pending);
```
I don't think the callback will ever be called.

Personally, I don't think it should be called either. Not sure what you think?

jhedberg replied:
It depends on what guarantees the API documentation gives for the callback. At the very least that documentation should be updated to mention that it's possible that the callback never comes. What's the current behaviour if you disconnect a connection while a callback is pending?

alwa-nordic replied:
We should invoke all callbacks unless we have a good reason not to do so. Dropping callbacks in some circumstances complicates the API.

sem1-nordic replied:
Will remove `sys_slist_init(&conn->tx_pending);` and the callbacks will be dealt with in `conn_tx_destroy` in `conn_cleanup`

sem1-nordic commented:
@Thalley Just rebased and ran locally and still seem to be hitting the same assert unfortunately

Thalley commented:
> @Thalley Just rebased and ran locally and still seem to be hitting the same assert unfortunately

Do we have an issue raised for it? If not, then I suggest to create one. 

One thing that concerns me, is that the test is basically a copy of the existing GATT tests, right? So why is it failing on the reset test, and not in the regular test? Wouldn't that indicate an issue when resetting the BT system?

Thalley commented:
> > > @Thalley Just rebased and ran locally and still seem to be hitting the same assert unfortunately
> > 
> > 
> > Do we have an issue raised for it? If not, then I suggest to create one.
> > One thing that concerns me, is that the test is basically a copy of the existing GATT tests, right? So why is it failing on the reset test, and not in the regular test? Wouldn't that indicate an issue when resetting the BT system?
> 
> Ok I will raise an issue.
> 
> I think it is a bug with the hci_driver_close implementation in the zephyr link layer.

Possibly. I am not convinced your changes are working correctly without a test case to provide some evidence. We should either wait for that issue to be resolved, or the current test case should be extended to at least do a BT connection.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1737,6 +1742,10 @@ int bt_unpair(uint8_t id, const bt_addr_le_t *addr)
 		return 0;
 	}
 
+	if (!addr) {
+		addr = BT_ADDR_LE_ANY;
+	}
```
What does this do? 

From I can tell, `unpair` has no handling for for `BT_ADDR_LE_ANY` so I am not convinced this change makes any sense in the current state.

sem1-nordic replied:
Changed to not run unpair with BT_ADDR_LE_ANY or NULL

sem1-nordic commented:
> > > Do we have an issue raised for it? If not, then I suggest to create one.
> > > One thing that concerns me, is that the test is basically a copy of the existing GATT tests, right? So why is it failing on the reset test, and not in the regular test? Wouldn't that indicate an issue when resetting the BT system?
> > 
> > 
> > Ok I will raise an issue.
> > I think it is a bug with the hci_driver_close implementation in the zephyr link layer.
> 
> Possibly. I am not convinced your changes are working correctly without a test case to provide some evidence. We should either wait for that issue to be resolved, or the current test case should be extended to at least do a BT connection.
I do have target tests running with NCS and the softdevice controller for these cases:
1. enable, scan, disable
2. enable, adv, disable
3. enable, connect, disable
4. enable, connect, l2cap connect, l2cap send, gatt write, encrypt, l2cap send, gatt write, disconnect, disable

Thalley commented:
> > > > Do we have an issue raised for it? If not, then I suggest to create one.
> > > > One thing that concerns me, is that the test is basically a copy of the existing GATT tests, right? So why is it failing on the reset test, and not in the regular test? Wouldn't that indicate an issue when resetting the BT system?
> > > 
> > > 
> > > Ok I will raise an issue.
> > > I think it is a bug with the hci_driver_close implementation in the zephyr link layer.
> > 
> > 
> > Possibly. I am not convinced your changes are working correctly without a test case to provide some evidence. We should either wait for that issue to be resolved, or the current test case should be extended to at least do a BT connection.
> 
> I do have target tests running with NCS and the softdevice controller for these cases:
> 
>     1. enable, scan, disable
> 
>     2. enable, adv, disable
> 
>     3. enable, connect, disable
> 
>     4. enable, connect, l2cap connect, l2cap send, gatt write, encrypt, l2cap send, gatt write, disconnect, disable

So there is only an issue with reset HCI with the Zephyr controller? At least that would indicate that the host supports the re-enable changes here :)

sem1-nordic commented:
> > > Possibly. I am not convinced your changes are working correctly without a test case to provide some evidence. We should either wait for that issue to be resolved, or the current test case should be extended to at least do a BT connection.
> > 
> > 
> > I do have target tests running with NCS and the softdevice controller for these cases:
> > ```
> > 1. enable, scan, disable
> > 
> > 2. enable, adv, disable
> > 
> > 3. enable, connect, disable
> > 
> > 4. enable, connect, l2cap connect, l2cap send, gatt write, encrypt, l2cap send, gatt write, disconnect, disable
> > ```
> 
> So there is only an issue with reset HCI with the Zephyr controller? At least that would indicate that the host supports the re-enable changes here :)

Indeed, I believe so.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1731,13 +1736,15 @@ int bt_unpair(uint8_t id, const bt_addr_le_t *addr)
+	if (!addr || bt_addr_le_eq(addr, BT_ADDR_LE_ANY)) {
+		if (IS_ENABLED(CONFIG_BT_SMP)) {
+			bt_foreach_bond(id, unpair_remote, &id);
+			return 0;
```
```suggestion
```

Thalley reviewed (changes requested):
Minor comment, otherwise LGTM

sem1-nordic commented:
Fix was merged for the zephyr assert, so have added the disable_with_gatt test back in.

jhedberg commented on subsys/bluetooth/host/adv.c:
```diff
@@ -269,6 +269,17 @@ void bt_le_ext_adv_foreach(void (*func)(struct bt_le_ext_adv *adv, void *data),
+void bt_adv_reset_adv_pool(void)
+{
+#if defined(CONFIG_BT_EXT_ADV)
+	for (size_t i = 0; i < ARRAY_SIZE(adv_pool); i++) {
+		(void)memset(&adv_pool[i], 0, sizeof(adv_pool[i]));
+	}
```
Can't this be done with a single `memset(adv_pool, 0, sizeof(adv_pool))`?

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -699,6 +699,29 @@ static void conn_cleanup(struct bt_conn *conn)
+
+	bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+
+	/* bt_conn_unref is called at the end of bt_conn_foreach,
+	 * we set to 1 here, so it will be 0 at the end of
+	 * bt_conn_foreach.
+	 */
+	atomic_set(&conn->ref, 1);
```
This looks very suspicious and likely wrong. The reference count should only ever be manipulated using the `bt_conn_unref()` and `bt_conn_ref()` functions, and those functions should be the only place that change the count directly using `atomic_*()`.

sem1-nordic replied:
Dealing with the tx_pending callbacks meant this was no longer needed, so have removed

jhedberg commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -103,6 +103,21 @@ int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
 	return 0;
 }
 
+void bt_pub_key_hci_disrupted(void)
```
Let's use something else than "disrupted" here. I'd propose "hci_reset" since that's what it's used for.

Thalley replied:
Can't this also be disrupted if there is a disconnect between the key being requested, and the key being received?

alwa-nordic replied:
Naming it "HCI reset" suggest that it's called in for every HCI reset message, including the one in bt_enable.

jhedberg replied:
> Can't this also be disrupted if there is an ACL disconnect between the key being requested, and the key being received?

No, this is not related at to connections. The key pair creation and dhkey calculation functions are done independent of any connections.



> Naming it "HCI reset" suggest that it's called in for every HCI reset message, including the one that's part of bt_enable.

Sure, if you want 100% correctness we can call the API on bt_enable() as well, I'd be fine with it. Note that we're talking about an internal API however, not something that would be visible to general users of the public Bluetooth API. If you have a better name than "hci_reset" for what happens when we reset HCI, but not when we reset HCI for the first time, I'm open to suggestions :)

sem1-nordic replied:
I think disrupted is better personally.

alwa-nordic replied:
> If you have a better name than "hci_reset" for what happens when we reset HCI, but not when we reset HCI for the first time, I'm open to suggestions :)

"Disrupted" or something similar is good I think. Maybe "HCI link lost"? "HCI uninitialized"? "HCI down"? (Inspired by ifconfig eth0 up/down.).

There may be a long time between the "link is lost", and HCI reset. The relevant event is when the HCI state machine re-enters the uninitialized / start state. That can later transition to the "enabling" state, triggering the actual HCI reset message.

sean-madigan committed:
Bluetooth: Host: clean up after connections in bt_disable

Reset disconnected handles as is no longer needed after disable.

Created a destroy function to clear all connection states and cancel any
deferred work.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3780,10 +3787,29 @@ int bt_disable(void)
 
 	bt_monitor_send(BT_MONITOR_CLOSE_INDEX, NULL, 0);
 
+#if defined(CONFIG_BT_EXT_ADV) || defined(CONFIG_BT_PERIPHERAL)
```
shouldn't `PERIPHERAL` rather be `CONFIG_BT_BROADCASTER`?

jori-nordic commented on subsys/bluetooth/host/id.c:
```diff
@@ -1368,6 +1364,32 @@ int bt_id_delete(uint8_t id)
+
+int bt_id_delete_all(void)
+{
+	for (uint8_t id = 0U; id < bt_dev.id_count; id++) {
```
shouldn't we cache `bt_dev.id_count`? It seems it is decremented in `id_delete()` so I'm worried the flow of this for-loop might be erratic.

sem1-nordic replied:
Yeah that's a good point

alwa-nordic commented on subsys/bluetooth/host/adv.c:
```diff
@@ -269,6 +269,15 @@ void bt_le_ext_adv_foreach(void (*func)(struct bt_le_ext_adv *adv, void *data),
+{
+#if defined(CONFIG_BT_EXT_ADV)
+	(void)memset(&adv_pool, 0, sizeof(adv_pool));
+#else
+	(void)memset(&bt_dev.adv, 0, sizeof(bt_dev.adv));
+#endif /* defined(CONFIG_BT_EXT_ADV) */
```
I think we want to also zero out `bt_dev.adv` when `CONFIG_BT_EXT_ADV=y`. I suspect `adv_get_legacy` will malfunction otherwise.

```suggestion
#endif /* defined(CONFIG_BT_EXT_ADV) */
	(void)memset(&bt_dev.adv, 0, sizeof(bt_dev.adv));
```

sean-madigan committed:
Bluetooth: Host: reset adv_pool in bt_disable

Reset the adv_pool in disable so that advertising can be started after
re-enabling without issue.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3780,10 +3787,29 @@ int bt_disable(void)
+#if defined(CONFIG_BT_CONN)
+#if defined(CONFIG_BT_SMP)
+	bt_pub_key_hci_disrupted();
+#endif /* CONFIG_BT_SMP */
```
Please use `if (IS_ENABLED(CONFIG_...) {` whenever you can. In this case it should be possible

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3780,10 +3787,29 @@ int bt_disable(void)
+	disconnected_handles_reset();
+#endif /* CONFIG_BT_CONN */
+
+	err = bt_id_delete_all();
```
This doesn't look correct to me. It's possible to create new identities even before calling `bt_enable()` the first time, so why should `bt_disable()` touch them?

sem1-nordic replied:
How about a function `bt_id_reset_all()`?

jhedberg replied:
But why do we want to do anything about identities to begin with? `bt_disable()` is intended to just shut down the controller and reset any HCI-related host-side states and resources. We're not clearing stuff like the GATT database, unpairing devices, etc, so why would we be resetting or removing local identities?

sem1-nordic replied:
I see, will remove

sean-madigan committed:
Bluetooth: Host: add pub_key_hci_disrupted function

Add bt_pub_key_hci_disrupted function for when hci is disrupted during
a pub key hci command, to clear callbacks and BT_DEV_PUB_KEY_BUSY flag.

Call this from bt_disable for the case bt_disable disrupts the pub key
hci command.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>
